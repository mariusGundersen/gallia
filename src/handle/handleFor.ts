import { insertAfter } from "../domUtils.js";
import listDiff from "../listDiff.js";
import {
  makeObservable,
  ObservableObject,
  ObservationScope,
} from "../observable.js";
import { makeExpressionEvaluator, makeKeyEvaluator } from "../utils.js";
import { CreateWalker, HandleGenerator } from "./index.js";

interface Source {
  key: string;
  index: number;
  value: unknown;
  startOfItem: Node;
  endOfItem: Node;
  observable: {
    value: unknown;
    index: number;
  };
  destroy: () => void;
}

export function* handleFor(
  element: HTMLTemplateElement,
  createWalker: CreateWalker
): HandleGenerator {
  const forExpression = element.getAttribute("x-for");
  if (!forExpression) return;

  const [isFor, name, collection] =
    /^(\w+)\s+of\s+(.*)$/.exec(forExpression) ?? [];
  if (!isFor) return;

  const expression = makeExpressionEvaluator(collection);
  const getKey = getKeyEvaluator(name, element);

  const documentFragment = element.content;
  documentFragment.prepend(
    documentFragment.ownerDocument.createComment(`start of item`)
  );
  documentFragment.append(
    documentFragment.ownerDocument.createComment(`end of item`)
  );

  const walk = createWalker(documentFragment);

  yield (node: Node, data: ObservableObject, scope: ObservationScope) => {
    const element = node as HTMLTemplateElement;
    const parent = element.parentNode;
    const before: Node = element.ownerDocument.createComment(
      `for (${forExpression}) {`
    );
    const after: Node = element.ownerDocument.createComment("}");
    parent!.insertBefore(before, element);
    parent!.replaceChild(after, element);

    let oldList: { key: string; value: unknown }[] = [];
    let oldMap = new Map<string, number>();
    const keyedValues = new Map<string, Source>();
    scope.observeAndReact(
      () => expression(data) as unknown[],
      (items) => {
        let index = 0;
        let endOfPreviousItem = before;
        const currentList = items.map((value, index) => ({
          key: getKey(value, index),
          value,
        }));

        oldMap = listDiff(oldList, oldMap, currentList, {
          insert({ key, value }) {
            const clone = documentFragment.cloneNode(true);
            const startOfItem = clone.firstChild as Node;
            startOfItem.textContent = `start of ${index} with key ${key}`;
            const endOfItem = clone.lastChild as Node;
            endOfItem.textContent = `end of ${index}`;

            const { scope: subScope, destroy } = scope.createSubScope();

            const source: Source = {
              key,
              value,
              index,
              startOfItem: startOfItem,
              endOfItem: endOfItem,
              destroy: destroy,
              observable: makeObservable({
                value,
                index,
              }),
            };

            keyedValues.set(key, source);

            const subData = createSubData(data, name, source);

            // remember the childNodes so that we can send them to the handlers
            const fragment = rememberTheChildren(clone);

            insertAfter(endOfPreviousItem, clone);

            walk(fragment, subData, subScope);

            endOfPreviousItem = endOfItem;
            index++;
          },
          move(item) {
            const source = keyedValues.get(item.key) as Source;
            source.index = index;
            source.observable.index = index;
            if (source.value !== item.value) {
              source.value = item.value;
              source.observable.value = item.value;
            }

            moveItemAfter(
              source.startOfItem,
              source.endOfItem,
              endOfPreviousItem
            );

            endOfPreviousItem = source.endOfItem;
            index++;
          },
          noop(item) {
            // item has not moved
            const source = keyedValues.get(item.key) as Source;
            if (source.value !== item.value) {
              source.value = item.value;
              source.observable.value = item.value;
            }

            endOfPreviousItem = source.endOfItem;
            index++;
          },
          remove(item) {
            const source = keyedValues.get(item.key) as Source;
            source.destroy();
            removeItem(source.startOfItem, source.endOfItem);
            keyedValues.delete(item.key);
          },
        });

        oldList = currentList;
      }
    );
  };
}

const indexKeyExpression = (_: unknown, $index: number) => $index;

function createSubData(data: ObservableObject, name: string, source: Source) {
  return Object.create(data, {
    [name]: {
      get() {
        return source.observable.value;
      },
    },
    $index: {
      get() {
        return source.observable.index;
      },
    },
    $parent: {
      get() {
        return data;
      },
    },
  });
}

function rememberTheChildren(clone: Node): Node {
  return {
    childNodes: Array.from(clone.childNodes),
  } as any;
}

function getKeyEvaluator(itemName: string, elm: HTMLTemplateElement) {
  const keyExpression = elm.getAttribute("x-key");

  return keyExpression
    ? makeKeyEvaluator(itemName, keyExpression)
    : indexKeyExpression;
}

function moveItemAfter(
  startOfItem: Node,
  endOfItem: Node,
  endOfPreviousItem: Node
) {
  if (endOfPreviousItem.nextSibling === startOfItem) return;
  let item = startOfItem;
  while (true) {
    const nextItem = item.nextSibling;
    insertAfter(endOfPreviousItem, item);
    endOfPreviousItem = item;
    if (item === endOfItem) break;
    if (!nextItem) break;
    item = nextItem;
  }
}

function removeItem(startOfItem: Node, endOfItem: Node) {
  let item = startOfItem;
  while (true) {
    const nextItem = item.nextSibling;
    item.parentNode!.removeChild(item);
    if (item === endOfItem) break;
    if (!nextItem) break;
    item = nextItem;
  }
}
