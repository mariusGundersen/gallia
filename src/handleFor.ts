import { insertAfter } from "./domUtils";
import {
  makeObservable,
  ObservableObject,
  ObservationScope,
} from "./observable";
import { makeExpressionEvaluator } from "./utils";

interface Source {
  key: unknown;
  index: number;
  value: unknown;
  startOfItem: Node;
  endOfItem: Node;
  destroy: () => void;
}

export function handleFor(
  element: HTMLTemplateElement,
  data: ObservableObject,
  scope: ObservationScope,
  walk?: (node: Node, data: ObservableObject, scope: ObservationScope) => void
) {
  const forExpression = element.getAttribute("x-for");
  if (!forExpression) return;

  const [isFor, name, collection] =
    /^(\w+)\s+of\s+(.*)$/.exec(forExpression) ?? [];
  if (!isFor) return;

  const expression = makeExpressionEvaluator(collection);
  const parent = element.parentNode;
  const before: Node = element.ownerDocument.createComment(
    `for (${forExpression}) {`
  );
  const after: Node = element.ownerDocument.createComment("}");
  parent!.insertBefore(before, element);
  parent!.replaceChild(after, element);

  const keyExpression = getKeyExpression(element);

  const oldItems = new Map<unknown, Source>();
  scope.observeAndReact(
    () => expression(data),
    (items) => {
      const oldList = [...oldItems.values()].sort((a, b) => a.index - b.index);
      let endOfPreviousItem = before;
      let index = 0;
      for (const value of items) {
        const key = keyExpression({ [name]: value, $index: index });
        const oldItem = oldItems.get(key);
        if (oldItem?.index === index) {
          console.log("item has not moved", index, value);
          // item has not moved
          oldList.splice(oldList.indexOf(oldItem), 1);
          endOfPreviousItem = oldItem.endOfItem;
          if (oldItem.value !== value) {
            oldItem.value = value;
          }
        } else if (!oldItem) {
          console.log("new item", index, value);
          // this is a new item
          // insert it at the current position
          const clone = element.content.cloneNode(true);
          const startOfItem = element.ownerDocument.createComment(
            `start of ${index} with key ${key}`
          );
          const endOfItem = element.ownerDocument.createComment(
            `end of ${index}`
          );

          const source = ({
            key,
            value,
            index,
            startOfItem,
            endOfItem,
          } as unknown) as Source;
          const subData = Object.create(
            data,
            makeObservable({
              [name]: {
                get() {
                  return source.value;
                },
              },
              $index: {
                get() {
                  return source.index;
                },
              },
            })
          );
          const { scope: subScope, destroy } = scope.createSubScope();
          source.destroy = destroy;
          walk?.(clone, subData, subScope);
          clone.appendChild(endOfItem);
          clone.insertBefore(startOfItem, clone.firstChild);
          insertAfter(endOfPreviousItem, clone);
          endOfPreviousItem = endOfItem;
          oldItems.set(key, source);
        } else {
          console.log("item has moved from", oldItem.index, "to", index, value);
          // item has moved
          // move dom elements
          oldItem.index = index;
          if (oldItem.value !== value) {
            oldItem.value = value;
          }

          let item = oldItem.startOfItem;
          do {
            const nextItem = item.nextSibling;
            insertAfter(endOfPreviousItem, item);
            endOfPreviousItem = item;
            if (!nextItem) break;
            item = nextItem;
          } while (item !== oldItem.endOfItem);

          oldList.splice(oldList.indexOf(oldItem), 1);
        }

        index++;
      }

      // remove any remaining items from before
      for (const oldItem of oldList) {
        console.log("item has been removed", oldItem.index);
        oldItems.delete(oldItem.key);
        oldItem.destroy();
        let item = oldItem.startOfItem;
        do {
          const nextItem = item.nextSibling;
          item.parentNode!.removeChild(item);
          if (!nextItem) break;
          item = nextItem;
        } while (item !== oldItem.endOfItem);
      }
    }
  );
}

const indexKeyExpression = ({ $index }: { $index: number }) => $index;

function getKeyExpression(elm: HTMLTemplateElement) {
  const key = elm.getAttribute("x-key");

  return key ? makeExpressionEvaluator(key) : indexKeyExpression;
}
