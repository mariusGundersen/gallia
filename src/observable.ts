export interface Observer {
  notify(): void;
  targets: TargetPropTuple[];
}
type Listener = () => void;

export type ObservableObject = Record<string, unknown> | any[];

type Key = string | number | symbol;

type TargetPropTuple = [ObservableObject, Key];

class Observed {
  observers = new Set<Observer>();
  notify() {
    // Prevent infinite loop
    //
    // .values() returns an iterator
    // calling add() will replace the item already in the list
    // Therefore calling add(existingItem) while iterating will
    // result in an endless loop, as the exsitingItem is picked
    // up again.
    const observers = Array.from(this.observers.values());
    for (let i = 0; i < observers.length; i++) {
      observers[i].notify();
    }
  }
  add(listener: Observer) {
    this.observers.add(listener);
  }
  remove(listener: Observer) {
    this.observers.delete(listener);
  }
}

// This is the global single current observer
// There can only be one observer at a time
let currentObserver: Observer | undefined;

// Here we store both the direct and reverse relationship between an observer and the observed
// observer -> object -> property
// object[propeprty] -> observer
// They are weakmaps so that we don't leak memory
const subscriptions = new WeakMap<ObservableObject, Map<Key, Observed>>();

function clearObserver(observer: Observer) {
  while (observer.targets.length > 0) {
    const [target, prop] = observer.targets.pop() as TargetPropTuple;
    // remove the subscription that would notify this observer
    subscriptions.get(target)?.get(prop)?.remove(observer);
  }
}

/**
 * Take an object and return an observable version of the object
 *
 * @param object The object we want to observe
 */
export function makeObservable<T extends ObservableObject>(object: T): T {
  // prevent an already observable object being made observable again
  if (subscriptions.has(object)) return object;

  // All the properties on the object can be observed
  const listeners = new Map<Key, Observed>();

  const getOrCreate = (key: Key) => {
    const found = listeners.get(key);
    if (found) return found;

    const created = new Observed();
    listeners.set(key, created);
    return created;
  };

  const result = new Proxy(object, {
    get(target, prop, receiver) {
      if (currentObserver && prop !== Symbol.unscopables) {
        getOrCreate(prop).add(currentObserver);
        currentObserver.targets.push([receiver, prop]);
      }
      return Reflect.get(target, prop, receiver);
    },
    set(target, prop, value, receiver) {
      const result = Reflect.set(target, prop, value, receiver);
      if (prop !== Symbol.unscopables) {
        getOrCreate(prop).notify();
      }
      return result;
    },
    deleteProperty(target, prop) {
      listeners.delete(prop);
      return Reflect.deleteProperty(target, prop);
    },
  });
  subscriptions.set(result, listeners);
  return result;
}

/**
 * Extend this class to become observable
 */
export class Observable {
  constructor() {
    return makeObservable(this as ObservableObject);
  }
}

function observe<T>(observer: Observer, getter: () => T) {
  clearObserver(observer);
  currentObserver = observer;
  try {
    return getter();
  } finally {
    // finally ensures that it works even with exceptions
    currentObserver = undefined;
  }
}

function unobserve(observer: Observer) {
  // Forget everything we know about this observer
  clearObserver(observer);
}

interface Destroyable {
  addDestroyListener(listener: Listener): void;
  removeDestroyListener(listener: Listener): void;
}

interface Context extends Destroyable {
  observe<T>(observer: Observer, getter: () => T): T;
  unobserve(observer: Observer): void;
}

const rootContext: Context = {
  observe,
  unobserve,
  addDestroyListener() {},
  removeDestroyListener() {},
};

export interface ObservationScope {
  observeAndReact<T>(getter: () => T, reaction: (value: T) => void): () => void;
  onDestroy(listener: Listener): void;
  createSubScope(): { scope: ObservationScope; destroy(): void };
}

function createObservationScope(context: Context): ObservationScope {
  return {
    observeAndReact<T>(getter: () => T, reaction: (value: T) => void) {
      const observer: Observer = {
        notify() {
          reaction(context.observe(observer, getter));
        },
        targets: [],
      };

      observer.notify();

      // if there is nothing that can notify this observer then it's not needed
      // unobserve removes it from the list in the context
      if (observer.targets.length === 0) {
        context.unobserve(observer);
        return noop;
      }

      return () => context.unobserve(observer);
    },
    onDestroy(listener: Listener) {
      context.addDestroyListener(listener);
    },
    createSubScope() {
      const [destroy, subContext] = createSubContext(context);
      return {
        scope: createObservationScope(subContext),
        destroy,
      };
    },
  };
}

const globalObservationScope = createObservationScope(rootContext);

export { globalObservationScope };

function noop() {}

function createSubContext({
  addDestroyListener,
  removeDestroyListener,
}: Destroyable): [() => void, Context] {
  let observers: Set<Observer> | null = new Set<Observer>();
  const destroyListeners = new Set<Listener>();

  function destroy() {
    if (!observers)
      throw new Error("Attempted to destory an already destroyed scope");

    removeDestroyListener(destroy);

    for (const observer of observers) {
      unobserve(observer);
    }

    for (const destroyListener of destroyListeners) {
      destroyListener();
    }

    observers = null;
  }

  addDestroyListener(destroy);

  const subContext: Context = {
    observe<T>(observer: Observer, effect: () => T) {
      if (!observers) throw new Error("Attempted to observe a destroyed scope");

      observers.add(observer);
      return observe(observer, effect);
    },
    unobserve(observer: Observer) {
      if (!observers)
        throw new Error("Attempted to unobserve a destroyed scope");

      observers.delete(observer);
      unobserve(observer);
    },
    addDestroyListener(listener: Listener) {
      destroyListeners.add(listener);
    },
    removeDestroyListener(listener: Listener) {
      destroyListeners.delete(listener);
    },
  };

  return [destroy, subContext];
}
