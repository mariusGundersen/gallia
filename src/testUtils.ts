export function createElementFromHTML(htmlString: string) {
  var div = document.createElement("div");
  div.innerHTML = htmlString.trim();
  return div.firstChild as HTMLElement;
}
