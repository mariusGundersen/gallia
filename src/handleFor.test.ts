import { handleFor } from "./handleFor";
import { globalObservationScope } from "./observable";
import { createElementFromHTML } from "./testUtils";

describe("for", () => {
  test("empty list", () => {
    const parent = createElementFromHTML(`
      <div>
        <template x-for="item of list">
          Something
        </template>
      </div>
    `);

    const data = {
      list: [],
    };

    const spy = jest.fn();

    handleFor(
      parent.firstElementChild as HTMLTemplateElement,
      data,
      globalObservationScope,
      spy
    );

    expect(parent.innerHTML).toMatchSnapshot();

    expect(spy).not.toHaveBeenCalled();
  });

  test("one item", () => {
    const parent = createElementFromHTML(`
      <div>
        <template x-for="item of list">
          Something
        </template>
      </div>
    `);

    const data = {
      list: [1],
    };

    const spy = jest.fn();

    handleFor(
      parent.firstElementChild as HTMLTemplateElement,
      data,
      globalObservationScope,
      spy
    );

    expect(parent.innerHTML).toMatchSnapshot();

    expect(spy).toHaveBeenCalledTimes(1);
    expect(spy.mock.calls[0][1].list).toBe(data.list);
    expect(spy.mock.calls[0][1].item).toBe(1);
    expect(spy.mock.calls[0][1].$index).toBe(0);
  });

  test("five items", () => {
    const parent = createElementFromHTML(`
      <div>
        <template x-for="item of list">
          Something
        </template>
      </div>
    `);

    const data = {
      list: [1, 2, 3, 4, 5],
    };

    const spy = jest.fn();

    handleFor(
      parent.firstElementChild as HTMLTemplateElement,
      data,
      globalObservationScope,
      spy
    );

    expect(parent.innerHTML).toMatchSnapshot();

    expect(spy).toHaveBeenCalledTimes(5);
  });

  test("iterator with five items", () => {
    const parent = createElementFromHTML(`
      <div>
        <template x-for="item of list()">
          Something
        </template>
      </div>
    `);

    const data = {
      *list() {
        for (let i = 0; i < 5; i++) {
          yield i;
        }
      },
    };

    handleFor(
      parent.firstElementChild as HTMLTemplateElement,
      data,
      globalObservationScope
    );

    expect(parent.innerHTML).toMatchSnapshot();
  });
});
