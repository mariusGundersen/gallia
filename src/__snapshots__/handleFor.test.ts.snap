// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`for empty list 1`] = `
"
        <!--for (item of list) {--><!--}-->
      "
`;

exports[`for five items 1`] = `
"
        <!--for (item of list) {--><!--start of 0 with key 0-->
          Something
        <!--end of 0--><!--start of 1 with key 1-->
          Something
        <!--end of 1--><!--start of 2 with key 2-->
          Something
        <!--end of 2--><!--start of 3 with key 3-->
          Something
        <!--end of 3--><!--start of 4 with key 4-->
          Something
        <!--end of 4--><!--}-->
      "
`;

exports[`for iterator with five items 1`] = `
"
        <!--for (item of list()) {--><!--start of 0 with key 0-->
          Something
        <!--end of 0--><!--start of 1 with key 1-->
          Something
        <!--end of 1--><!--start of 2 with key 2-->
          Something
        <!--end of 2--><!--start of 3 with key 3-->
          Something
        <!--end of 3--><!--start of 4 with key 4-->
          Something
        <!--end of 4--><!--}-->
      "
`;

exports[`for one item 1`] = `
"
        <!--for (item of list) {--><!--start of 0 with key 0-->
          Something
        <!--end of 0--><!--}-->
      "
`;
